File: web/src/reduce.test.ts
// Cross-language fixture parity: the TypeScript reducer must produce the same
// State the Go reducer does over the very same internal/state/testdata/*.json
// fixtures. An expected state may omit an interaction map or the submit/close
// signals that stay at their zero value; normalizeInteractions fills them, the
// same way the Go test's initMaps does, and a JSON round-trip drops the
// undefined optionals (note, summary) the way Go's omitempty tags do.

import { readFileSync, readdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { applyEvent, applyInteraction, reduce } from './reduce';
import type { Doc } from './schema';
import type { Interactions, PresentEvent, PresentState, Revising, Rounds } from './events';

const testdataDir = fileURLToPath(new URL('../../internal/state/testdata/', import.meta.url));

interface Fixture {
  name: string;
  events: PresentEvent[];
  expected: { doc: Doc; interactions?: Partial<Interactions>; rounds?: Partial<Rounds>; revising?: Revising };
}

function normalizeInteractions(i: Partial<Interactions> | undefined): Interactions {
  return {
    decisions: i?.decisions ?? {},
    choices: i?.choices ?? {},
    inputs: i?.inputs ?? {},
    packs: i?.packs ?? {},
    feedback: i?.feedback ?? {},
    replies: i?.replies ?? {},
    annotations: i?.annotations ?? {},
    triage: i?.triage ?? {},
    submitted: i?.submitted ?? { value: false, revision: 0 },
    closed: i?.closed ?? { value: false },
  };
}

// normalizeRounds fills a round record's `packs`, `annotations`, and `triage`
// maps the same way Go's initMaps does (reduce_test.go), so a fixture that omits
// an empty snapshot still matches the reducer, which always emits them.
function normalizeRounds(r: Partial<Rounds> | undefined): Rounds {
  return {
    current: r?.current ?? 1,
    currentTitle: r?.currentTitle,
    blockRounds: r?.blockRounds ?? {},
    history: (r?.history ?? []).map((rec) => ({
      ...rec,
      packs: rec.packs ?? {},
      annotations: rec.annotations ?? {},
      triage: rec.triage ?? {},
    })),
  };
}

// normalizeRevising fills the empty working set a fixture omits, the same way the
// Go test's initMaps normalizes State.Revising.
function normalizeRevising(r: Revising | undefined): Revising {
  return { blockIds: r?.blockIds ?? [], note: r?.note };
}

function canonical(state: PresentState): unknown {
  return JSON.parse(JSON.stringify(state)) as unknown;
}

function loadFixtures(): Fixture[] {
  return readdirSync(testdataDir)
    .filter((f) => f.endsWith('.json'))
    .sort()
    .map((f) => JSON.parse(readFileSync(testdataDir + f, 'utf8')) as Fixture);
}

describe('reduce fixture parity', () => {
  const fixtures = loadFixtures();

  it('finds the shared fixtures', () => {
    expect(fixtures.length).toBeGreaterThan(0);
  });

  for (const fx of fixtures) {
    it(fx.name, () => {
      const got = reduce(fx.events);
      const want: PresentState = {
        doc: fx.expected.doc,
        interactions: normalizeInteractions(fx.expected.interactions),
        rounds: normalizeRounds(fx.expected.rounds),
        revising: normalizeRevising(fx.expected.revising),
      };
      expect(canonical(got)).toEqual(canonical(want));
    });
  }
});

describe('applyInteraction optimistic path', () => {
  const base = reduce([
    { origin: 'agent', type: 'block.upserted', seq: 1, payload: { block: { id: 'a1', type: 'approval' } } },
  ]);

  it('applies a decision optimistically before the echo', () => {
    const next = applyInteraction(base, { type: 'decision.created', blockId: 'a1', verdict: 'approved' });
    expect(next.interactions.decisions.a1).toEqual({ verdict: 'approved' });
  });

  it('carries an other write-in through the optimistic choice echo', () => {
    const next = applyInteraction(base, { type: 'choice.selected', blockId: 'ch1', optionIds: [], other: 'custom' });
    expect(next.interactions.choices.ch1).toEqual({ optionIds: [], other: 'custom' });
  });

  it('defers submit to the echo, leaving the cache unchanged', () => {
    // A failed POST must not strand the UI past the round; the optimistic patch
    // is a no-op and the SSE echo advances the round instead.
    expect(applyInteraction(base, { type: 'submit', revision: 3 })).toBe(base);
  });

  it('advances the round when the submit event itself replays', () => {
    const echoed = applyEvent(base, { origin: 'human', type: 'submit', seq: 2, payload: { revision: 3 } });
    expect(echoed.rounds.current).toBe(2);
    expect(echoed.rounds.history.map((r) => r.submittedRevision)).toEqual([3]);
  });
});

describe('reduce errors', () => {
  it('rejects an unknown event type', () => {
    const events = [
      { origin: 'agent', type: 'bogus.event', seq: 1, payload: {} },
    ] as unknown as PresentEvent[];
    expect(() => reduce(events)).toThrow(/unknown event type/);
  });

  it('rejects an invalid verdict', () => {
    const events = [
      { origin: 'human', type: 'decision.created', seq: 1, payload: { blockId: 'a1', verdict: 'maybe' } },
    ] as unknown as PresentEvent[];
    expect(() => reduce(events)).toThrow(/invalid verdict/);
  });
});

File: internal/state/reduce_test.go
package state_test

import (
	"embed"
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	"github.com/yasyf/cc-present/internal/doc"
	"github.com/yasyf/cc-present/internal/state"
)

//go:embed testdata/*.json
var fixturesFS embed.FS

type fixture struct {
	Name     string          `json:"name"`
	Events   []state.Event   `json:"events"`
	Expected json.RawMessage `json:"expected"`
}

// TestFixtures drives the reducer entirely from the language-neutral JSON
// fixtures in testdata; Phase 2's TypeScript reducer consumes the same files.
// An expected state may omit an interaction map that stays empty: initMaps
// normalizes it to the empty map the reducer always produces.
func TestFixtures(t *testing.T) {
	entries, err := fixturesFS.ReadDir("testdata")
	if err != nil {
		t.Fatalf("read testdata: %v", err)
	}
	if len(entries) == 0 {
		t.Fatal("no fixtures found in testdata")
	}
	for _, entry := range entries {
		name := "testdata/" + entry.Name()
		data, err := fixturesFS.ReadFile(name)
		if err != nil {
			t.Fatalf("read %s: %v", name, err)
		}
		var fx fixture
		if err := json.Unmarshal(data, &fx); err != nil {
			t.Fatalf("unmarshal fixture %s: %v", name, err)
		}
		if fx.Name == "" {
			t.Fatalf("fixture %s has no name", name)
		}
		t.Run(fx.Name, func(t *testing.T) {
			got, err := state.Reduce(fx.Events)
			if err != nil {
				t.Fatalf("Reduce() error = %v", err)
			}
			var want state.State
			if err := json.Unmarshal(fx.Expected, &want); err != nil {
				t.Fatalf("unmarshal expected: %v", err)
			}
			initMaps(&want)
			assertStateEqual(t, got, want)
		})
	}
}

func TestReduceErrors(t *testing.T) {
	tests := []struct {
		name    string
		events  []state.Event
		wantErr string
	}{
		{
			name:    "unknown event type",
			events:  []state.Event{{Origin: "agent", Type: "bogus.event", Seq: 1, Payload: []byte(`{}`)}},
			wantErr: "unknown event type",
		},
		{
			name:    "invalid verdict",
			events:  []state.Event{{Origin: "human", Type: "decision.created", Seq: 1, Payload: []byte(`{"blockId":"a1","verdict":"maybe"}`)}},
			wantErr: "invalid verdict",
		},
		{
			name:    "malformed block on upsert",
			events:  []state.Event{{Origin: "agent", Type: "block.upserted", Seq: 1, Payload: []byte(`{"block":{"id":"x","type":"bogus"}}`)}},
			wantErr: "unknown type",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := state.Reduce(tt.events)
			if err == nil {
				t.Fatalf("Reduce() error = nil, want substring %q", tt.wantErr)
			}
			if !strings.Contains(err.Error(), tt.wantErr) {
				t.Fatalf("Reduce() error = %q, want substring %q", err.Error(), tt.wantErr)
			}
		})
	}
}

// TestReducePackInteraction covers the pack.interaction reducer case with inline
// events; the language-neutral testdata fixtures for it land in Phase 3 lockstep
// with the TypeScript reducer.
func TestReducePackInteraction(t *testing.T) {
	const packDoc = `{"doc":{"version":1,"title":"T","blocks":[{"id":"r1","type":"example.rating","value":3}]},"revision":1}`

	t.Run("last write wins per block", func(t *testing.T) {
		st, err := state.Reduce([]state.Event{
			{Origin: "agent", Type: "doc.replaced", Seq: 1, Payload: []byte(packDoc)},
			{Origin: "human", Type: "pack.interaction", Seq: 2, Payload: []byte(`{"blockId":"r1","payload":{"value":3}}`)},
			{Origin: "human", Type: "pack.interaction", Seq: 3, Payload: []byte(`{"blockId":"r1","payload":{"value":5}}`)},
		})
		if err != nil {
			t.Fatalf("Reduce: %v", err)
		}
		if got := packValue(t, st.Interactions.Packs["r1"]); got != 5 {
			t.Fatalf("packs[r1].value = %d, want 5 (last write wins)", got)
		}
	})

	t.Run("snapshot into the closed round on submit", func(t *testing.T) {
		st, err := state.Reduce([]state.Event{
			{Origin: "agent", Type: "doc.replaced", Seq: 1, Payload: []byte(packDoc)},
			{Origin: "human", Type: "pack.interaction", Seq: 2, Payload: []byte(`{"blockId":"r1","payload":{"value":4}}`)},
			{Origin: "human", Type: "submit", Seq: 3, Payload: []byte(`{"revision":1}`)},
		})
		if err != nil {
			t.Fatalf("Reduce: %v", err)
		}
		if len(st.Rounds.History) != 1 {
			t.Fatalf("history = %d rounds, want 1", len(st.Rounds.History))
		}
		if got := packValue(t, st.Rounds.History[0].Packs["r1"]); got != 4 {
			t.Fatalf("round packs[r1].value = %d, want 4", got)
		}
		// The live interaction persists past the snapshot, mirroring choices/decisions.
		if got := packValue(t, st.Interactions.Packs["r1"]); got != 4 {
			t.Fatalf("live packs[r1].value = %d, want 4", got)
		}
	})
}

func packValue(t *testing.T, v state.PackValue) int {
	t.Helper()
	var p struct {
		Value int `json:"value"`
	}
	if err := json.Unmarshal(v.Payload, &p); err != nil {
		t.Fatalf("decode pack payload %q: %v", v.Payload, err)
	}
	return p.Value
}

func initMaps(s *state.State) {
	if s.Interactions.Decisions == nil {
		s.Interactions.Decisions = map[string]state.Decision{}
	}
	if s.Interactions.Choices == nil {
		s.Interactions.Choices = map[string]state.Selection{}
	}
	if s.Interactions.Inputs == nil {
		s.Interactions.Inputs = map[string]state.InputValue{}
	}
	if s.Interactions.Packs == nil {
		s.Interactions.Packs = map[string]state.PackValue{}
	}
	if s.Interactions.Feedback == nil {
		s.Interactions.Feedback = map[string][]state.Feedback{}
	}
	if s.Interactions.Replies == nil {
		s.Interactions.Replies = map[string][]state.Reply{}
	}
	if s.Doc == nil {
		s.Doc = &doc.Doc{Version: 1, Blocks: []doc.Block{}}
	}
	if s.Rounds.Current == 0 {
		s.Rounds.Current = 1
	}
	if s.Rounds.BlockRounds == nil {
		s.Rounds.BlockRounds = map[string]int{}
	}
	if s.Rounds.History == nil {
		s.Rounds.History = []state.RoundRecord{}
	}
	if s.Revising.BlockIDs == nil {
		s.Revising.BlockIDs = []string{}
	}
	for i := range s.Rounds.History {
		if s.Rounds.History[i].Packs == nil {
			s.Rounds.History[i].Packs = map[string]state.PackValue{}
		}
	}
}

func assertStateEqual(t *testing.T, got, want state.State) {
	t.Helper()
	gotJSON, err := json.MarshalIndent(got, "", "  ")
	if err != nil {
		t.Fatalf("marshal got: %v", err)
	}
	wantJSON, err := json.MarshalIndent(want, "", "  ")
	if err != nil {
		t.Fatalf("marshal want: %v", err)
	}
	var gotAny, wantAny any
	if err := json.Unmarshal(gotJSON, &gotAny); err != nil {
		t.Fatalf("re-unmarshal got: %v", err)
	}
	if err := json.Unmarshal(wantJSON, &wantAny); err != nil {
		t.Fatalf("re-unmarshal want: %v", err)
	}
	if !reflect.DeepEqual(gotAny, wantAny) {
		t.Errorf("state mismatch\n--- got ---\n%s\n--- want ---\n%s", gotJSON, wantJSON)
	}
}
